src/UriSetHostCommon.c

```diff
@@ -0,0 +1,258 @@
+					}
+				}
+				break;
+			case URI_HOST_TYPE_IPFUTURE:
+				{
+					const int res = URI_FUNC(IsWellFormedHostIpFutureMm)(first, afterLast, memory);
+					assert((res == URI_SUCCESS) || (res == URI_ERROR_SYNTAX) || (res == URI_ERROR_MALLOC));
+					if (res != URI_SUCCESS) {
+						return res;
+					}
+				}
+				break;
+			case URI_HOST_TYPE_REGNAME:
+				if (URI_FUNC(IsWellFormedHostRegName)(first, afterLast) == URI_FALSE) {
+					return URI_ERROR_SYNTAX;
+				}
+				break;
+			default:
+				assert(0 && "Unsupported URI host type");
+		}
+	}
+
+	{
+		/* Clear old value */
+		const UriBool hadHostBefore = URI_FUNC(HasHost)(uri);
+		if (uri->hostData.ipFuture.first != NULL) {
+			if ((uri->owner == URI_TRUE) && (uri->hostData.ipFuture.first != uri->hostData.ipFuture.afterLast)) {
+				memory->free(memory, (URI_CHAR *)uri->hostData.ipFuture.first);
+			}
+			uri->hostData.ipFuture.first = NULL;
+			uri->hostData.ipFuture.afterLast = NULL;
+			uri->hostText.first = NULL;
+			uri->hostText.afterLast = NULL;
+		} else if (uri->hostText.first != NULL) {
+			if ((uri->owner == URI_TRUE) && (uri->hostText.first != uri->hostText.afterLast)) {
+				memory->free(memory, (URI_CHAR *)uri->hostText.first);
+			}
+			uri->hostText.first = NULL;
+			uri->hostText.afterLast = NULL;
+		}
+
+		if (uri->hostData.ip4 != NULL) {
+			memory->free(memory, uri->hostData.ip4);
+			uri->hostData.ip4 = NULL;
+		} else if (uri->hostData.ip6 != NULL) {
+			memory->free(memory, uri->hostData.ip6);
+			uri->hostData.ip6 = NULL;
+		}
+
+		/* Already done setting? */
+		if (first == NULL) {
+			/* Yes, but disambiguate as needed */
+			if (hadHostBefore == URI_TRUE) {
+				uri->absolutePath = URI_TRUE;
+
+				{
+					const UriBool success = URI_FUNC(EnsureThatPathIsNotMistakenForHost)(uri, memory);
+					return (success == URI_TRUE)
+							? URI_SUCCESS
+							: URI_ERROR_MALLOC;
+				}
+			}
+
+			return URI_SUCCESS;
+		}
+	}
+
+	assert(first != NULL);
+
+	/* Ensure owned */
+	if (uri->owner == URI_FALSE) {
+		const int res = URI_FUNC(MakeOwnerMm)(uri, memory);
+		if (res != URI_SUCCESS) {
+			return res;
+		}
+	}
+
+	assert(uri->owner == URI_TRUE);
+
+	/* Apply new value; NOTE that .hostText is set for all four host types */
+	{
+		URI_TYPE(TextRange) sourceRange;
+		sourceRange.first = first;
+		sourceRange.afterLast = afterLast;
+
+		if (URI_FUNC(CopyRangeAsNeeded)(&uri->hostText, &sourceRange, memory) == URI_FALSE) {
+			return URI_ERROR_MALLOC;
+		}
+
+		uri->absolutePath = URI_FALSE;  /* always URI_FALSE for URIs with host  */
+
+		/* Fill .hostData as needed */
+		switch (hostType) {
+			case URI_HOST_TYPE_IP4:
+				{
+					uri->hostData.ip4 = memory->malloc(memory, sizeof(UriIp4));
+					if (uri->hostData.ip4 == NULL) {
+						return URI_ERROR_MALLOC;
+					}
+
+					{
+						const int res = URI_FUNC(ParseIpFourAddress)(uri->hostData.ip4->data, first, afterLast);
+						assert(res == URI_SUCCESS);  /* because checked for well-formedness earlier */
```

Comment:
I'd like to note that this causes a compiler warning if `NDEBUG` is example, since `res` is unused. Please see:

https://github.com/php/php-src/pull/19636/commits/1a5d94d762eb1bc8ab2318d622707bac194d9e50

The workaround in that commit doesn't seem quite right, though. A `(void)res;` seems to work:


```suggestion
						const int res = URI_FUNC(ParseIpFourAddress)(uri->hostData.ip4->data, first, afterLast);
						(void)res; /* suppress unused variable warning if assertions are disabled */
						assert(res == URI_SUCCESS);  /* because checked for well-formedness earlier */
```

Reply:
@TimWolla that's an interesting one! For one, because it raises the question whether the uriparser upstream CI matrix should cover the `NDEBUG` case. For two, it makes me wonder why PHP is compiling with fatal warnings when it is third party code. With compilers getting stricter by the minute (for good and for bad) that seems like asking for future trouble. What do you think? PS: I'll add a variant of the `(void)res;` with the next push…

Reply:
> For two, it makes me wonder why PHP is compiling with fatal warnings when it is third party code. With compilers getting stricter by the minute (for good and for bad) that seems like asking for future trouble. What do you think?

The build system of PHP is a complex autoconf-based beast. I can't claim to understand it :smiley: Building everything with `-Werror` in CI makes sense to me, since (expected) warnings from vendored libraries make it harder to review the CI logs in case of legit warnings for PHP's own code. In case of legit warnings that are caught by PHP's CI, they should be fixed upstream and in other cases, PHP is likely going to carry a patch.

For reference, the relevant config.m4 is this one: https://github.com/php/php-src/blob/master/ext/uri/config.m4

Reply:
This thread is resolved (unless you have follow-up questions).

Reply:
@TimWolla I'll resist going too much more into that topic to not derail the review, but would like to be clear that for (both libexpat and) uriparser I am happy to address warnings upstream *but* I'll want bug reports rather than patches for compiler warnings in the future please, and address them myself.